src/main/scala/com/microsoft/hyperspace/index/zordercovering/ZOrderCoveringIndexConfig.scala

```diff
@@ -0,0 +1,60 @@
+/*
+ * Copyright (2021) The Hyperspace Project Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.microsoft.hyperspace.index.zordercovering
+
+import org.apache.spark.sql.DataFrame
+
+import com.microsoft.hyperspace.Hyperspace
+import com.microsoft.hyperspace.index._
+import com.microsoft.hyperspace.index.covering.{CoveringIndex, CoveringIndexConfigTrait}
+import com.microsoft.hyperspace.util.HyperspaceConf
+
+/**
+ * ZOrderCoveringIndexConfig specifies the configuration of a covering index which is z-ordered.
+ *
+ * Use this class to create a Z-ordered covering index with [[Hyperspace.createIndex()]].
+ *
+ * @param indexName Index name.
+ * @param indexedColumns Columns to be used for calculating z-address.
+ * @param includedColumns Columns to be included in the index.
+ */
+case class ZOrderCoveringIndexConfig(
+    override val indexName: String,
+    override val indexedColumns: Seq[String],
+    override val includedColumns: Seq[String] = Seq())
+    extends CoveringIndexConfigTrait {
+  override def createIndex(
+      ctx: IndexerContext,
+      sourceData: DataFrame,
+      properties: Map[String, String]): (ZOrderCoveringIndex, DataFrame) = {
+    val (indexData, resolvedIndexedColumns, resolvedIncludedColumns) =
+      CoveringIndex.createIndexData(
+        ctx,
+        sourceData,
+        indexedColumns,
+        includedColumns,
+        IndexUtils.hasLineageColumn(properties))
+
+    val index = ZOrderCoveringIndex(
+      resolvedIndexedColumns.map(_.normalizedName),
+      resolvedIncludedColumns.map(_.normalizedName),
+      indexData.schema,
+      HyperspaceConf.ZOrderCovering.approxSourceBytesPerPartition(ctx.spark),
```

Comment:
Just thinking out loud.

At first, I thought using `numBucketsForIndex` for the number of index files would be better for simplicity and consistency, as the difference between hash bucketed covering index and Z-order covering index is how index files are ordered and everything else can remain the same.

But that will make the number of index files kind of fixed to 200 files, as many users won't care about the numBuckets config. It can be an issue if the data is very large and 200 files are too few for efficient handling. But at the same time, it's also an issue for hash bucketed covering index. So we either rely on the user to change the config appropriately for their data or need to find a better way.

With the current code, small data (< 10 GB) won't get much boost with Z-order index, as there will be too few index files.

Anyway, it would be nice if we decide on a way to determine the number of files, whether it is the old way (numBuckets) or new (target size per file), and apply it consistently for both hash bucketed covering index and Z-order covering index.

Reply:
Ah, I forgot about the importance of exact numbers for bucketing. Then I think it's okay to have separate parameters.